### ISSUE TITLE
property policies

### ISSUE THREAD

**Issue by RWOverdijk** [balderdashy/sails#2496]
Hey,

I've made a custom implementation for this, but I thought I'd run it by you guys to see if perhaps there's interest to implement it in sails.

## The concept
The concept is pretty simple. Just as with regular policies, you have policies for properties. These policies are parsed in a rule-set to make sure there's no performance loss. The rules are defined in `config/propertyPolicies.js`.

### Usage
In it's most basic form, the config could look something like this:

**config/propertyPolicies.js:**
```js
module.exports.propertyPolicies = {
  user: ['email', 'mailable'] // Whitelist for model `user`
};
```

This specifies that the _update_ blueprint is allowed to only update these two properties. Sending along id, password or any of the other properties will result in those values simply being ignored.

To achieve this, I've created a policy that does these checks for you, and populates a blacklist. Like so:

**api/policies/filterProperties.js:**
```js
var config = sails.config.propertyPolicies;

/**
 * Apply policies to models as to manage access to properties.
 *
 * @todo Allow for property filtering in find
 * @todo Allow short-hand (model: ['whitelist'])
 * @todo Allow blacklisting ({blacklist: ['blacklist']}). Use Object.getOwnProperties(sails.models[model].definition); to extract properties.
 * @todo Allow Default (*) On global, and per-model basis like with regular policies
 * @todo Allow applying policies (or even specific dir with modelPolicies?)
 */
module.exports = function(req, res, next) {
  var model = [req.options.model],
      blacklist = [];

  if (!config[model]) {
    return next();
  }

  Object.getOwnPropertyNames(sails.models[model].definition).forEach(function (property) {
    if (config[model].indexOf(property) > -1) {
      return;
    }

    blacklist.push(property);
  });

  if (!req.options.values) {
    req.options.values = {};
  }

  req.options.values.blacklist = blacklist;

  next();
};
```

Then, in `config/policies.js` just add `'filterProperties'`.

### Filtering properties
Currently this concept only applies to update. However, in the same train of thought I figured that setting up policies like this for _read_ should be possible to. This could be achieved by setting up the same config structure as policies.js currently allows, except one level deeper:

**config/propertyPolicies.js:**
```js
module.exports.propertyPolicies = {
  // Default policy for all models and properties.
  '*': false,
  User: {
    create: {
      '*'  : true,
      roles: 'isAdmin'
    },
    update: {
      mailable: 'ownsRecord',
      email   : 'ownsRecord',
      roles   : 'isAdmin'
    },
    find  : {
      email   : ['isauthenticated', 'ownsRecord'],
      username: 'isAuthenticated',
      id      : 'isAuthenticated'
    }
  }
};
```

This allows for conditional properties based on rules. Obviously limiting properties is possible through toJSON already, but that's like specifying '*' for a specific model; it's applied to every scenario.

Finally, I'd like to point out this has nothing to do with limiting the number of properties returned through the blueprints. Obviously you _could_ create such a propertyPolicy and do it yourself. But afaik waterline2 already allows you to specify which properties you want to fetch.

**config/propertyPolicies.js:**
```
module.exports.policies = {

  // Default policy for all models and properties.
  '*': false,

  User: {
    find  : {
      '*': 'limitProperties'
    }
};
```

**api/propertyPolicies/limitProperties.js (untested):**

```js
module.exports = function (req, property, next) {
  if (!req.param('properties')) {
    return next(null, true); // True. You're allowed to use this property.
  }

  var properties = req.param('properties');

  return next(null, properties.indexOf(property) > -1);
};
```

## Conclusion
In conclusion, this allows you to keep it as simple or complex as you like. I think this does give you a lot of control.

Thoughts? Feedback? Let me know.

**Comment by Dallas62** [comment]
:+1: 
Will try it ;)

**Comment by loicsaintroch** [comment]
I like it. Would you mind writing a PR a get some community feedback about this improvement? I'll be glad to merge it. Thank you @RWOverdijk, you rock!
